### PR TITLE
Migrate 'components/claim' module and associated tests to TypeScript

### PR DESCRIPTION
**Pull request by egrinberg**
#### Summary

Migrate 'components/claim' module and associated tests to TypeScript

#### Ticket Link

Fixes https://github.com/mattermost/mattermost-server/issues/16144

### CODE REVIEW

File: package-lock.json [hunk 1]
```diff
@@ -3847,9 +3847,9 @@
           }
         },
         "node-fetch": {
-          "version": "2.6.0",
-          "resolved": "https://registry.npmjs.org/node-fetch/-/node-fetch-2.6.0.tgz",
-          "integrity": "sha512-8dG4H5ujfvFiqDmVu9fQ5bOHUC15JMjMY/Zumv26oOvvVJjM67KF8koCWIabKQ1GJIa9r2mMZscBq/TbdOcmNA==",
+          "version": "2.6.1",
```
Comment (hanzei) [inline package-lock.json:3850]: Could you please remove these unrelated changes?

**Review by hanzei — COMMENTED** [review]
Thanks for submitting this PR :tada: One request before we kick of reviews.

File: components/claim/claim_controller.tsx [hunk 2]
```diff
@@ -12,26 +12,33 @@ import EmailToOAuth from 'components/claim/components/email_to_oauth';
...
+type Props = {
+    location: Location;
+    siteName?: string;
+    ldapLoginFieldName?: string;
+    passwordConfig: PasswordConfig;
+    match: {
+        url: string
+    },
+    actions: {
+        switchLdapToEmail: (ldapPassword: string, email: string, emailPassword: string, mfaCode: string | undefined) => ActionFunc;
```
Comment (Willyfrog) [inline components/claim/claim_controller.tsx:36]:
```suggestion
        switchLdapToEmail: (ldapPassword: string, email: string, emailPassword: string, mfaCode?: string) => ActionFunc;
```

File: components/claim/claim_controller.tsx [hunk 3]
```diff
@@ -12,26 +12,33 @@ import EmailToOAuth from 'components/claim/components/email_to_oauth';
...
 
-        actions: PropTypes.shape({
-            switchLdapToEmail: PropTypes.func.isRequired,
-        }).isRequired,
-    };
+type Props = {
+    location: Location;
+    siteName?: string;
+    ldapLoginFieldName?: string;
+    passwordConfig: PasswordConfig;
```
Comment (Willyfrog) [inline components/claim/claim_controller.tsx:31]: from the previous code, I'm guessing the `passwordconfig` prop is not mandatory
```suggestion
    passwordConfig?: PasswordConfig;
```

**Review by Willyfrog — CHANGES_REQUESTED** [review]
looking good, but some minor changes, thanks!

File: package-lock.json (hunk 1 above)
Comment (egrinberg) [reply package-lock.json]: Yes sure, fixed in the latest commit.

File: components/claim/claim_controller.tsx (hunk 2 above)
Comment (egrinberg) [reply components/claim/claim_controller.tsx]: thank you, changed in the latest commit.

File: components/claim/claim_controller.tsx (hunk 3 above)
Comment (egrinberg) [reply components/claim/claim_controller.tsx]: done

**Review by Willyfrog — APPROVED** [review]
LGTM! thank you!!

**Review by nevyangelova — APPROVED** [review]

**Review by jgilliam17 — APPROVED** [review]
Thank you @egrinberg 
Tested, looks good to merge.
- Verified setting up account around third party auth. 
QA will test further after merge as some sys. console settings around LDAP are set to true by default through env. variable.
